adb/adb_protocol.py

```diff
@@ -409,3 +429,128 @@ def StreamingCommand(cls, usb, service, command='', timeout_ms=None):
         timeout_ms=timeout_ms)
     for data in connection.ReadUntilClose():
       yield data.decode('utf8')
+
+  @classmethod
+  def InteractiveShellCommand(cls, conn, cmd=None, strip_cmd=True, delim=None, strip_delim=True, clean_stdout=True):
+    """Retrieves stdout of the current InteractiveShell and sends a shell command if provided
+    TODO: Should we turn this into a yield based function so we can stream all output?
```

Comment:
StreamingCommand was supposed to provide this, why all this logic for delimiters and such?

It seems you're trying to use the InteractiveShellCommand non-interactively? If you meant to use it interactively, why not just keep the connection open and just keep writing the user's input into it? I haven't tried this myself, but I'm pretty sure you can just keep the connection open and keep writing to it.

Also, I just realized there's a shell v2!!! https://github.com/aosp-mirror/platform_system_core/blob/9a38df3f5f03d1ed20bf9a196a53ec00eb5970e0/adb/commandline.cpp#L265

I'm not sure what version of android/ADB supports it, but it started showing up in AOSP in 2015, so likely Nougat and later had it. The main benefit of v2 is that it returns the exit code, so we can finally return (exit_code, stdout, stderr) when supported!

Okay, obviously don't worry about that here, I was just excited when I noticed that comment :)

Reply:
'Interactive' is a bit of a misnomer, now that I think about it. The use case here is that I have programs that need the shell's state to persist across calls, which requires the connection be left open. For example, if I set a shell variable with either Shell() or StreamingShell(), that variable is gone in the next call since ADB tears down that shell's instance.

InteractiveShell() is a larger function because it is serving two purposes:
1. Being able to guarantee persistence of a shell's state over multiple calls to an ADB device - ie. Set some shell variables, run a program that may switch users on a rooted device, do work, then switch users back to the 'shell' user --all of which isn't possible without holding that connection open

2. Affording the above guarantees to a user that is 'interacting' with the shell (ie. add stuff to my PATH, run some programs)


Because of the additional automation / consumption by non-humans that is being placed around InteractiveShell(), I've added those output 'cleaning' functions to strip out the interpreter-added characters. This is required for cases like:
output = self.device.InteractiveShell("/path/to/program args")
do_stuff_with_program_results(output)

If the interpreters added guff is stuffed in the output, it breaks some expectations of what the output 'should be' in the event this ADB shell isn't the source of that output (ie. simple subprocess stdout return won't have interpreter characters like "shell@hammerhead #" in there)

Hope that made sense!

Reply:
The shell v2 stuff looks pretty cool actually, would be a good thing to work on outside of this :D